rpc/src/module/indexer.rs

```diff
@@ -0,0 +1,121 @@
+use ckb_indexer::IndexerStore;
+use jsonrpc_core::Result;
+use jsonrpc_derive::rpc;
+use jsonrpc_types::{BlockNumber, CellTransaction, LiveCell, LockHashIndexState, Unsigned};
+use numext_fixed_hash::H256;
+
+#[rpc]
+pub trait IndexerRpc {
+    #[rpc(name = "get_live_cells_by_lock_hash")]
+    fn get_live_cells_by_lock_hash(
+        &self,
+        _lock_hash: H256,
+        _page: Unsigned,
+        _per_page: Unsigned,
+        _reverse_order: Option<bool>,
+    ) -> Result<Vec<LiveCell>>;
+
+    #[rpc(name = "get_transactions_by_lock_hash")]
+    fn get_transactions_by_lock_hash(
+        &self,
+        _lock_hash: H256,
+        _page: Unsigned,
+        _per_page: Unsigned,
+        _reverse_order: Option<bool>,
+    ) -> Result<Vec<CellTransaction>>;
+
+    #[rpc(name = "index_lock_hash")]
+    fn index_lock_hash(
+        &self,
+        _lock_hash: H256,
+        _index_from: Option<BlockNumber>,
+    ) -> Result<LockHashIndexState>;
+
+    #[rpc(name = "deindex_lock_hash")]
+    fn deindex_lock_hash(&self, _lock_hash: H256) -> Result<()>;
+
+    #[rpc(name = "get_lock_hash_index_states")]
+    fn get_lock_hash_index_states(&self) -> Result<Vec<LockHashIndexState>>;
+}
+
+pub(crate) struct IndexerRpcImpl<WS> {
+    pub store: WS,
+}
+
+impl<WS: IndexerStore + 'static> IndexerRpc for IndexerRpcImpl<WS> {
+    fn get_live_cells_by_lock_hash(
+        &self,
+        lock_hash: H256,
+        page: Unsigned,
+        per_page: Unsigned,
+        reverse_order: Option<bool>,
+    ) -> Result<Vec<LiveCell>> {
+        let per_page = (per_page.0 as usize).min(50);
+        Ok(self
+            .store
+            .get_live_cells(
+                &lock_hash,
+                (page.0 as usize).saturating_mul(per_page),
+                per_page,
+                reverse_order.unwrap_or_default(),
+            )
+            .into_iter()
+            .map(Into::into)
+            .collect())
+    }
+
+    fn get_transactions_by_lock_hash(
+        &self,
+        lock_hash: H256,
+        page: Unsigned,
+        per_page: Unsigned,
+        reverse_order: Option<bool>,
+    ) -> Result<Vec<CellTransaction>> {
+        let per_page = (per_page.0 as usize).min(50);
+        Ok(self
+            .store
+            .get_transactions(
+                &lock_hash,
+                (page.0 as usize).saturating_mul(per_page),
+                per_page,
+                reverse_order.unwrap_or_default(),
+            )
+            .into_iter()
+            .map(Into::into)
+            .collect())
+    }
+
+    fn index_lock_hash(
```

Comment:
If I call this RPC repeatedly without removing the index first, the index may be in an invalid state.

For example:

- Assume that the current tip is 7. Call this RPC with `index_from` 5. Wait for the background thread to sync up to 7.
- Now tip becomes 10, and the background thread is still on 7.
- Call this RPC again with `index_from` 10.
- Background thread starts again, it will ignore blocks 8, 9

Reply:
it's rpc caller's responsibility to use an appropriate `index_from` value, indexer store won't handle this case.